README.adoc

```diff
@@ -0,0 +1,411 @@
+
+**Fail**
+
+```js
+const sum = () => {
+  const numbers = Array.prototype.slice.call(arguments)
+  return numbers.reduce((a, b) => a + b)
+}
+
+sum(1, 2, 3)
+```
+
+**Pass**
+
+```js
+const sum (numbers) => (
+  numbers.reduce((a, b) => a + b)
+)
+
+sum([1, 2, 3])
+
+const args = node.arguments
+```
+
+## `fp/no-class`
+
+> Classes are nice tools to use when programming with the object-oriented paradigm, as they hold internal state and give access to methods on the instances. In functional programming, having stateful objects is more harmful than helpful, and should be replaced by the use of pure functions.
+
+Further reading: https://github.com/joshburgess/not-awesome-es6-classes/[Not Awesome: ES6 Classes; A curated list of resources on why ES6 (aka ES2015) classes are NOT awesome]
+
+**Fail**
+
+```js
+class Polygon {
+  constructor (height, width) {
+    this.height = height
+    this.width = width
+  }
+}
+```
+
+**Pass**
+
+```js
+const polygon = (height, width) => ({
+  height: height,
+  width: width,
+})
+```
+
+## `fp/no-delete`
+
+> delete is an operator to remove fields from an object or elements from an array. This purposely mutates data, which is not wanted when doing functional programming.
+
+Further reading: https://github.com/google/google-api-nodejs-client/issues/375[Avoid using delete operator]
+
+**Fail**
+```js
+delete foo
+delete foo.bar
+delete foo[bar]
+```
+**Pass**
+```js
+import F from "ramda"
+
+const fooWithoutBar = F.omit(["bar"], foo)
+const fooWithoutField = F.omit([bar], foo)
+```
+
+## `fp/no-events`
+
+> The use of EventEmitter with the events module provided by Node.js promotes implicit side-effects by emitting and listening to events. Instead of events, you should prefer activating the wanted effects by calling the functions you wish to use explicitly.
+
+Probably what you should do is use a https://gist.github.com/staltz/868e7e9bc2a7b8c1f754[functional reactive programming] library: https://github.com/cujojs/most[`most`], https://github.com/Reactive-Extensions/RxJS[`rxjs`].
+
+**Fail**
+
+```js
+import EventEmitter from "events"
+```
+
+## `fp/no-get-set`
+
+**Fail**
+```js
+const person = {
+  name: 'Some Name',
+  get age () {
+    return this._age
+  },
+  set age (n) {
+    if (n < 0) {
+      this._age = 0
+    } else if (n > 100) {
+      this._age = 100
+    } else {
+      this._age = n
+    }
+  }: 20
+};
+
+person.__defineGetter__("name", function () {
+  return this.name || "John Doe";
+})
+
+person.__defineSetter__("name", function (name) {
+  this.name = name.trim();
+})
+```
+**Pass**
+```js
+import F from "ramda"
+
+const person = {
+  name: "Some Name",
+  age: 20,
+}
+
+const clamp = (n, min, max) => (
+  n <= min ? min :
+  n >= max ? max :
+             n
+)
+
+const setAge = (age, person) => (
+  F.merge(person, { age: clamp(age, 0, 100) })
+)
+```
+
+## `fp/no-let`
```

Comment:
I did some manual and automatic analysis of the `ui-website` codebase. Here's what I found.

### Stats

The stats show that `let` is only used for about 10% of assignments in source. That percentage jumps dramatic almost three-fold (28%) in tests. I did not manually analyze why that is for now. I did however inspect source. See next.

```
SOURCE

❯ ag 'const ' --stats-only --ignore ./tests
625 matches
118 files contained matches
186 files searched

❯ ag 'let ' --stats-only --ignore ./tests
82 matches
39 files contained matches
186 files searched


TESTS

❯ ag 'const ' --stats-only --ignore ./src
286 matches
58 files contained matches
77 files searched

❯ ag 'let ' --stats-only --ignore ./src
112 matches
51 files contained matches
77 files searched
```

### Manually Analyzed Cases

I went through almost every case of `let` in `source` (evidently some cases omitted in my table) to review the use-case. The refactor column contains the needed fix. Special values are `N/A` if its a valid use-case and `...` if I was unable to quickly review it (so potentially valid use-case too). I expect most `...` cases however would be invalid too. 

My overall findings were:
* valid instances 4-20% (very likely closer to lower bound)
* invalid instances 80-96% (very likely closer to upper bound)
* every single instance of `let` made code harder to reason about
* every single instance of `let` made code harder to parse (more verbose)
* except for a single strange optimization case every instance should eventually be able to be  refactored. `try-catch` cases would just need to use better libs with saner APIs that use Result types (e.g. [`Result Elm`](http://package.elm-lang.org/packages/elm-lang/core/latest/Result), [`Either PureScript`](https://pursuit.purescript.org/packages/purescript-either/1.0.0/docs/Data.Either), [`Either Haskell`](https://hackage.haskell.org/package/base-4.9.0.0/docs/Data-Either.html), [`Either JS`](https://github.com/folktale/data.either), [`Either Scala`](http://www.scala-lang.org/api/rc2/scala/Either.html)) instead of `throw`

Its worth noting here that I have already prepared an [ESLint refactor pass of `ms-search`](https://github.com/Groupe-Atallah/ms-search/pull/20) and the valid instances in that codebase amounted to 0% (again talking about `source`).

These findings help explain why we would want enforce `fp/no-let`. Its fine to use `let` if its truly needed, but almost always its not, and that's why we should be opting into it explicitly.

File |										reason |					refactor
----|----|----
config/_production.js |							file check, try block | 			…
config/index.js |								file check, if block |			…
src/client/cache-keys.js | 						cond string += |			push into array
src/client/mixins/truncate-text/index.js |			cond assign |			ternary
src/client/router/fetch-route.js |					cond string += |			URI lib || push into array
src/client/utils/cookie/index.js |					cond assign |			ternary
src/client/utils/cookie/index.js | for loop |					fold (aka reduce)
src/client/utils/cookie/index.js | switch |					functions
src/client/utils/debounce/index.js |				debounce util |				N/A
src/client/views/App.vue |						cond assign |			ternary
src/client/views/components/nav/Collapsible.vue |		loop+cond assign |		reduce
src/client/views/pages/BrandsList.vue |				Useless |					pick a different name 
src/client/views/pages/layout/header/LeftNav.vue |		cond string += |			pick a different name and/or ternary
src/client/views/pages/layout/header/RightNav.vue |		cond assign |			ternary
src/client/views/pages/layout/header/Search.vue |		cond string += |			pick a different name and/or ternary
src/client/views/pages/layout/mobile-nav/mixins/genderLinksMixin.js | cond string += | pick a different name and/or ternary
src/client/views/pages/layout/mobile-nav/mixins/genderLinksMixin.js | … | …
src/client/views/pages/layout/mobile-nav/SubMenu.vue | conditional assign | ternary
src/client/views/pages/product-listing/Description.vue | cond assign | ternary
src/client/views/pages/product-listing/left-nav/LeftNav.vue | cond assign | ternary
src/client/views/pages/product-listing/left-nav/MenuLevel.vue | cond assign | ternary
src/client/views/pages/product-listing/left-nav/MenuLevel.vue | loop+cond assign | reduce
src/client/views/pages/product-listing/Pagination.vue | cond string += | URI lib
src/client/views/pages/product-listing/mobile-refine/MobileRefine.vue | cond assign | ternary
src/client/views/pages/product-listing/ProductFilter.vue | loop+cond assign | reduce
src/client/views/pages/product-listing/ProductList.vue | cond assign | ternary
src/client/views/pages/product-listing/ProductSort.vue | loop+cond assign | reduce
src/server/App.ts | ? | ternary
src/server/controllers/product-groups/ProductGroupsController.ts | cond string += | URI lib
src/server/controllers/product-list/ProductListController.ts | … | …
src/server/helpers/String.ts | loop+cond assign | reduce and generally function pipeline
src/server/helpers/Request.ts | JSON parse, try block | Maybe type
src/server/helpers/Request.ts | request retrier | external lib
src/server/helpers/Request.ts | … | …
src/server/Logger.ts | switch | nested ternary
src/server/Logger.ts | … | …
src/server/middlewares/Internals.ts | Useless | const
src/server/middlewares/Localization.ts | cond assign | ternary
src/server/middlewares/Redirects.ts | cond assign | ternary
src/server/renderer/Renderer.ts | loop+cond string += | reduce
src/server/renderer/Renderer.ts | stream chunk counter | N/A or use FRP
src/server/renderer/Renderer.ts | optimization (apparently) | N/A or don’t do this…
src/server/renderer/RenderState.ts | useless | pick another name
src/server/services/SizeFilterService.ts | cond assign | ternary
src/store/getters.js | … | …
src/store/modules/mobile-nav.js | cond assign | nested ternary